File: SexyBiscuit.Engine/Gameplay/Pawn.cs
using Microsoft.Xna.Framework;
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Gameplay;

/// <summary>
/// An actor that a <see cref="Controller"/> can possess — the physical representation of a
/// player or an AI in the world. Modelled on Unreal's <c>APawn</c>.
/// </summary>
/// <remarks>
/// A pawn does not decide what it wants to do; its controller does. Controllers push intent
/// through <see cref="AddMovementInput"/> during Update, and the pawn drains that intent once
/// per frame in <see cref="ConsumeMovementInput"/>. Accumulating rather than assigning means
/// several input sources (stick, keyboard, an AI steering behaviour) compose without one
/// clobbering another.
/// </remarks>
public class Pawn : Actor
{
    /// <summary>The controller currently possessing this pawn, or null when unpossessed.</summary>
    public Controller? Controller { get; internal set; }

    /// <summary>True when the possessing controller is a <see cref="PlayerController"/>.</summary>
    public bool IsPlayerControlled => Controller is PlayerController;

    /// <summary>True when the pawn is possessed by any controller.</summary>
    public bool IsControlled => Controller != null;

    /// <summary>
    /// Rotation the controller wants the pawn to face, in degrees (pitch, yaw, roll).
    /// A first-person pawn typically applies yaw to itself and pitch to its camera.
    /// </summary>
    public Vector3 ControlRotation { get; set; }

    /// <summary>
    /// When true the pawn's yaw is driven from <see cref="ControlRotation"/> each frame.
    /// Turn this off for a pawn that should face its movement direction instead.
    /// </summary>
    public bool UseControllerRotationYaw { get; set; } = true;

    // Accumulated this frame, drained by ConsumeMovementInput.
    private Vector3 _pendingMovementInput;

    public Pawn() { }
    public Pawn(string name) : base(name) { }

    // -------------------------------------------------------------------------
    // Movement intent
    // -------------------------------------------------------------------------

    /// <summary>
    /// Adds movement intent in world space. Call from a controller during Update; the pawn's
    /// movement component drains it in FixedUpdate or LateUpdate.
    /// </summary>
    /// <param name="worldDirection">Direction to move. Need not be normalised.</param>
    /// <param name="scale">Multiplier, typically the analogue stick magnitude in 0..1.</param>
    public void AddMovementInput(Vector3 worldDirection, float scale = 1f)
        => _pendingMovementInput += worldDirection * scale;

    /// <summary>
    /// Returns the movement intent accumulated this frame and clears it. The returned vector
    /// is clamped to unit length so holding two directions is not faster than one.
    /// </summary>
    public Vector3 ConsumeMovementInput()
    {
        var v = _pendingMovementInput;
        _pendingMovementInput = Vector3.Zero;

        float lenSq = v.LengthSquared();
        return lenSq > 1f ? v / MathF.Sqrt(lenSq) : v;
    }

    /// <summary>Movement intent accumulated so far this frame, without clearing it.</summary>
    public Vector3 PendingMovementInput => _pendingMovementInput;

    /// <summary>Adds to the controller's yaw, in degrees.</summary>
    public void AddControllerYawInput(float degrees)
        => ControlRotation = ControlRotation with { Y = SBMath.WrapAngle(ControlRotation.Y + degrees) };

    /// <summary>Adds to the controller's pitch, in degrees, clamped to avoid gimbal flip.</summary>
    public void AddControllerPitchInput(float degrees, float minPitch = -89f, float maxPitch = 89f)
        => ControlRotation = ControlRotation with
        {
            X = MathHelper.Clamp(ControlRotation.X + degrees, minPitch, maxPitch)
        };

    // -------------------------------------------------------------------------
    // Possession callbacks
    // -------------------------------------------------------------------------

    /// <summary>Called after a controller takes control of this pawn.</summary>
    public virtual void OnPossessed(Controller controller) { }

    /// <summary>Called after a controller releases this pawn.</summary>
    public virtual void OnUnPossessed(Controller controller) { }

    /// <summary>
    /// Where the view should sit for a camera attached to this pawn, in world space.
    /// Override for an eye-height offset on a humanoid character.
    /// </summary>
    public virtual Vector3 GetViewLocation()
        => GetComponent<Transform3D>()?.Position ?? Vector3.Zero;
}

File: html5/src/gameplay/Pawn.js
// -----------------------------------------------------------------------------
// Pawn — an actor a Controller can possess.
// -----------------------------------------------------------------------------

import { Actor } from '../core/Actor.js';
import { registerActor } from '../core/TypeRegistry.js';
import { Transform3D } from '../core/Transform3D.js';
import { Vector3, SBMath } from '../math/index.js';

/** Something a player or an AI can drive. */
export class Pawn extends Actor {
    constructor(name = 'Pawn') {
        super(name);

        /** @type {?import('./Controller.js').Controller} */
        this.controller = null;

        /** Where the controller is looking: pitch, yaw and roll in degrees. */
        this.controlRotation = new Vector3(0, 0, 0);

        /** Turn the pawn to face the controller's yaw each frame. */
        this.useControllerRotationYaw = true;

        this.pendingMovementInput = new Vector3(0, 0, 0);

        if (!this.getComponent(Transform3D)) this.addComponent(Transform3D);
    }

    /** True when a PlayerController — rather than an AI — holds this pawn. */
    get isPlayerControlled() { return this.controller?.isPlayerController === true; }

    /** True when any controller holds this pawn. */
    get isControlled() { return this.controller != null; }

    /**
     * Adds a world-space movement request for this frame. Movement components
     * consume it; calling twice in a frame accumulates.
     */
    addMovementInput(worldDirection, scale = 1) {
        const d = Vector3.from(worldDirection);
        this.pendingMovementInput.x += d.x * scale;
        this.pendingMovementInput.y += d.y * scale;
        this.pendingMovementInput.z += d.z * scale;
    }

    /** Takes this frame's accumulated movement request and clears it. */
    consumeMovementInput() {
        const consumed = this.pendingMovementInput.clone();
        this.pendingMovementInput.set(0, 0, 0);
        return consumed;
    }

    /** Turns the controller's view. */
    addControllerYawInput(degrees) {
        this.controlRotation.y = SBMath.wrapAngle((this.controlRotation.y + degrees) * SBMath.DEG2RAD)
            * SBMath.RAD2DEG;
    }

    /** Pitches the controller's view, clamped so it cannot roll over the top. */
    addControllerPitchInput(degrees, minPitch = -89, maxPitch = 89) {
        this.controlRotation.x = SBMath.clamp(this.controlRotation.x + degrees, minPitch, maxPitch);
    }

    lateUpdate(_dt) {
        if (!this.useControllerRotationYaw) return;
        const t = this.transform3D;
        if (!t) return;

        // Only yaw: pitching the body as well would tip the whole character over
        // when the player looks up.
        const euler = t.eulerAngles;
        t.eulerAngles = new Vector3(euler.x, this.controlRotation.y, euler.z);
    }

    /** Where a camera attached to this pawn should sit. */
    getViewLocation() {
        return this.transform3D?.position ?? new Vector3(0, 0, 0);
    }

    /** Called when a controller takes this pawn. */
    onPossessed(controller) {}

    /** Called when a controller lets it go. */
    onUnPossessed(controller) {}
}
registerActor(Pawn, { category: 'Gameplay', summary: 'An actor a controller can possess.' });
